File: python/src/MCMC_epsDP.py
import numpy as np
from src.DPinR import DPinR

def MCMC_epsDP(N0, N1, X, Y, eps_DP, delta_DP, M, K, sigma_qe, sigma_qs, log_norm_var, s, sa, sb, est_s=1):
    """
    Implements a simplified MCMC-DP-Est algorithm.

    Args:
        N0, N1 (np.ndarray): Vectors of challenge numbers.
        X, Y (np.ndarray): Vectors of false positives and false negatives.
        eps_DP (float): Initial value for epsilon.
        delta_DP (float): Delta parameter of (epsilon, delta)-DP.
        M (int): Number of MCMC iterations.
        K (int): Number of auxiliary variables.
        sigma_qe, sigma_qs (float): Proposal standard deviations.
        log_norm_var (float): Variance for the log-normal prior of epsilon.
        s (float): Initial value for s.
        sa, sb (float): Shape parameters for the Beta prior of s.
        est_s (int): If 0, s is not estimated.

    Returns:
        tuple[np.ndarray, np.ndarray]: M samples for epsilon and s.
    """
    eps_DP_samps = np.zeros(M)
    s_samps = np.zeros(M)

    # Initialize
    AreaR_curr = 1 - 2 * (1 - delta_DP)**2 * np.exp(-eps_DP) / (1 + np.exp(-eps_DP))
    AreaR_curr_s = 1 - 2 * (1 - delta_DP * s)**2 * np.exp(-eps_DP * s) / (1 + np.exp(-eps_DP * s))
    n = len(X)
    a = 0.5 * np.ones(n)
    b = 0.5 * np.ones(n)

    for m in range(M):
        if (m + 1) % 10000 == 0:
            print(m + 1)
        
        # Propose eps_DP and s
        eps_DP_prop = np.exp(np.log(eps_DP) + sigma_qe * np.random.randn())
        s_prop = s + np.random.randn() * sigma_qs if est_s == 1 else s

        if 0 <= s_prop <= 1:
            # Log prior ratio
            if s == 0:
                log_prior_ratio = -(eps_DP_prop**2 - eps_DP**2) / log_norm_var
            else:
                log_prior_ratio = (-(eps_DP_prop**2 - eps_DP**2) / log_norm_var
                                   + (sa - 1) * np.log(s_prop) + (sb - 1) * np.log(1 - s_prop)
                                   - (sa - 1) * np.log(s) - (sb - 1) * np.log(1 - s))

            # Generate auxiliary variables
            A = np.vstack([a, np.random.rand(K - 1, n)])
            B = np.vstack([b, np.random.rand(K - 1, n)])

            # Calculate weights
            PYX = (N0 - X) * np.log(1 - A) + X * np.log(A) + (N1 - Y) * np.log(1 - B) + Y * np.log(B)

            InR_curr = DPinR(A, B, eps_DP, delta_DP) * (1 - DPinR(A, B, s * eps_DP, s * delta_DP))
            InR_prop = DPinR(A, B, eps_DP_prop, delta_DP) * (1 - DPinR(A, B, s_prop * eps_DP_prop, s_prop * delta_DP))
            
            AreaR_prop = 1 - 2 * (1 - delta_DP)**2 * np.exp(-eps_DP_prop) / (1 + np.exp(-eps_DP_prop))
            AreaR_prop_s = 1 - 2 * (1 - s_prop * delta_DP)**2 * np.exp(-s_prop * eps_DP_prop) / (1 + np.exp(-s_prop * eps_DP_prop))

            # Add a small constant to avoid log(0)
            log_W_prop = PYX + np.log(InR_prop.astype(float) + 1e-100) - np.log(AreaR_prop - AreaR_prop_s + 1e-100)
            log_W_curr = PYX + np.log(InR_curr.astype(float) + 1e-100) - np.log(AreaR_curr - AreaR_curr_s + 1e-100)

            # Acceptance rate (with log-sum-exp trick)
            max_curr = np.max(log_W_curr, axis=0)
            max_prop = np.max(log_W_prop, axis=0)
            log_w_numer_sum = np.log(np.sum(np.exp(log_W_prop - max_prop), axis=0)) + max_prop
            log_w_denom_sum = np.log(np.sum(np.exp(log_W_curr - max_curr), axis=0)) + max_curr
            log_r = np.sum(log_w_numer_sum) - np.sum(log_w_denom_sum) + log_prior_ratio
            
            if np.random.rand() < np.exp(log_r):
                eps_DP, s = eps_DP_prop, s_prop
                AreaR_curr, AreaR_curr_s = AreaR_prop, AreaR_prop_s
                W_temp = np.exp(log_W_prop - log_w_numer_sum)
            else:
                W_temp = np.exp(log_W_curr - log_w_denom_sum)

            # Sample (alpha, beta) pairs
            rand_uniform = np.random.rand(1, n)
            temp_ind_vec = np.sum(rand_uniform > np.cumsum(W_temp, axis=0), axis=0)
            a = A[temp_ind_vec, np.arange(n)]
            b = B[temp_ind_vec, np.arange(n)]

        # Store the sampled values
        eps_DP_samps[m] = eps_DP
        s_samps[m] = s

    return eps_DP_samps, s_samps

File: matlab/src/MCMC_epsDP.m
function [eps_DP_samps, s_samps] = MCMC_epsDP(N0, N1, X, Y, eps_DP, delta_DP, M, K, sigma_qe, sigma_qs, log_norm_var, s, sa, sb, est_s)

% [eps_DP_samps, s_samps] = MCMC_epsDP(N0, N1, X, Y, eps_DP, delta_DP, M, N, sigma_q, sigma_qs, log_norm_var, s, sa, sb, est_s)
%
% 
% This function implements MCMC-DP-Est in Algoritm 1 of the paper
% "MCMC for Bayesian estimation of Differential Privacy from Membership 
% Inference Attacks"
% 
% Inputs:
% N0: (1 x n) vector of numbers of challenges with theta ~ H0
% N1: (1 x n) vector of numbers of challenges with theta ~ H1
% 
% X: (1 x n) vector of false positives out of the N0 tests where H0 is true
% Y: (1 x n) vector of false negatives out of the N0 tests where H0 is true
% 
% eps_DP: Initial value for epsilon
% delta_DP: delta parameter of (epsilon, delta)-DP
% 
% M: Number of MCMC iterations
% K: Number of auxiliary variables (this MCMC algorithm is a MHAAR)
% 
% sigma_qe: standard deviation of the random walk proposal for epsilon
% sigma_qs: standard deviation of the random walk proposal for s
% 
% log_norm_var: The variance parameter of the log-normal prior for epsilon
%
% s: initial value for s
% sa, sb: The prior for s is Beta(sa, sb)
% est_s: Binary option. If set to 0, s is not estimated.
% 
% 
% Outputs:
% eps_DP_samps: M samples for epsilon
% s_samps: M samples s_samps

eps_DP_samps = zeros(1, M);
s_samps = zeros(1, M);

if nargin == 14
    est_s = 1;
end

% Initialize
AreaR_curr = 1 - 2 *(1 - delta_DP)^2*exp(-eps_DP)/(1 + exp(-eps_DP));
AreaR_curr_s = 1 - 2 *(1 - delta_DP*s)^2*exp(-eps_DP*s)/(1 + exp(-eps_DP*s));
n = length(X);
a = 0.5*ones(1, n);
b = 0.5*ones(1, n);

for m = 1:M
    if mod(m, 10000) == 0
        disp(m);
    end
    
    % Propose eps_DP and s
    eps_DP_prop = exp(log(eps_DP) + sigma_qe*randn);
    if est_s == 1
        s_prop = s + randn*sigma_qs;
    else
        s_prop = s;
    end

    % log prior ratio p(s', epsilon')/p(s, epsilon)
    if s == 0
        log_prior_ratio = -(eps_DP_prop^2 - eps_DP^2)/log_norm_var;
    else
        log_prior_ratio = -(eps_DP_prop^2 - eps_DP^2)/log_norm_var...
            + log(s_prop)*(sa-1) + log(1 - s_prop)*(sb-1)...
            - log(s)*(sa-1) - log(1 - s)*(sb-1);
    end

    if s_prop <= 1 && s_prop >= 0
        
        % generate samples for alpha_i and beta_i
        A = [a; rand(K-1, n)];
        B = [b; rand(K-1, n)];
    
        % Calculate the weights for the auxiliary variables
        PYX = (N0 - X).*log(1 - A) + X.*log(A) + (N1 - Y).*log(1 - B) + Y.*log(B);
    
        [InR_curr] = DPinR(A, B, eps_DP, delta_DP).*(1 - DPinR(A, B, s*eps_DP, s*delta_DP));
        [InR_prop] = DPinR(A, B, eps_DP_prop, delta_DP).*(1 - DPinR(A, B, s_prop*eps_DP_prop, s_prop*delta_DP));
    
        AreaR_prop = 1 - 2 *(1 - delta_DP)^2*exp(-eps_DP_prop)/(1 + exp(-eps_DP_prop));
        AreaR_prop_s = 1 - 2 *(1 - s*delta_DP)^2*exp(-s_prop*eps_DP_prop)/(1 + exp(-s_prop*eps_DP_prop));
    
        log_W_prop = PYX + log(double(InR_prop)) - log(AreaR_prop - AreaR_prop_s);
        log_W_curr = PYX + log(double(InR_curr)) - log(AreaR_curr - AreaR_curr_s);
    
        % acceptance rate
        max_curr = max(log_W_curr, [], 1);
        max_prop = max(log_W_prop, [], 1);
    
        log_w_numer_sum = log(sum(exp(log_W_prop - max_prop), 1)) + max_prop;
        log_w_denom_sum = log(sum(exp(log_W_curr - max_curr), 1)) + max_curr;
    
        log_numer = sum(log_w_numer_sum);
        log_denom = sum(log_w_denom_sum);
    
        log_r = log_numer - log_denom + log_prior_ratio;
    
        % decision
        if rand < exp(log_r)
            eps_DP = eps_DP_prop;
            s = s_prop;
            AreaR_curr = AreaR_prop;
            AreaR_curr_s = AreaR_prop_s;
            W_temp = exp(log_W_prop - log_w_numer_sum);
        else
            W_temp = exp(log_W_curr - log_w_denom_sum);
        end
    
        % sample the (alpha, beta) pairs
        temp_ind_vec = sum(rand(1, n) > cumsum(W_temp)) + 1;
        temp_mtx_ind_vec = K*(0:n-1) + temp_ind_vec;
        a = A(temp_mtx_ind_vec);
        b = B(temp_mtx_ind_vec);
    end

    % Store the sampled values
    eps_DP_samps(m) = eps_DP;
    s_samps(m) = s;

end
